core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -82,12 +107,28 @@ public CacheEventNotificationServiceImpl(ExecutorService orderedDelivery, Execut
   public void registerCacheEventListener(CacheEventListener<? super K, ? super V> listener,
```

Comment:
Objective – register and classify the listeners so as to maintain sync/async sets and enable threadlocal

What it internally does:
•	Listener will be eligible for registration iff it listens to atleast expirations or evictions
•	If listener is ordered, increament atomicInteger count
•	Pigeonhole listeners as sync/async and add it to the respective sets. Exception if already present
•	Finally, enable EventThreadLocal if not already enabled.

Why it should work:
Registering of a listener makes sure that EventThreadLocal gets enabled and adds to the orderedListener count so the eventDispatcher can be configured accordingly. Former is done under lock to prevent conflicting assignments. Maintaining sync/async listener sets helps in dispatch later.